src/main/java/picard/illumina/parser/readers/CbclReader.java

```diff
@@ -180,126 +258,147 @@ public void close() {
+            throw new PicardException(String.format("Error while decompressing from BCL file for cycle %d. Offending file on disk is %s",
                     (totalCycleCount + 1), this.streamFiles[totalCycleCount].getAbsolutePath()));
         }
 
         // Read uncompressed data from the buffer and expand each nibble into a full byte for ease of use
-        uncompressedByteBuffer.flip();
-        while (uncompressedByteBuffer.hasRemaining()) {
-            byte singleByte = uncompressedByteBuffer.get();
-            unNibbledByteBuffer.put((byte) (singleByte & 0x0f));
-            unNibbledByteBuffer.put((byte) ((singleByte >> 4) & 0x0f));
+        int index = 0;
+        for (byte singleByte : uncompressedByteArray) {
+            unNibbledByteArray[index] = (byte) (singleByte & 0x0f);
+            index++;
+            unNibbledByteArray[index] = (byte) ((singleByte >> 4) & 0x0f);
+            index++;
         }
         gzipInputStream.close();
 
         // Write buffer contents to cached tile array
-        unNibbledByteBuffer.flip();
-        //if nonPF reads are included we need to strip them out
+        // if nonPF reads are included we need to strip them out
         if (!currentCycleData.pfExcluded) {
-            ByteBuffer uncompressedFilteredByteBuffer = ByteBuffer.allocate(tileData.uncompressedBlockSize * 2);
-            FilterFileReader reader = new FilterFileReader(filters.get(tileData.tileNum));
-            while (reader.hasNext()) {
-                byte readByte = unNibbledByteBuffer.get();
-                if (reader.next()) {
-                    uncompressedFilteredByteBuffer.put(readByte);
+            boolean[] filterDatas = cachedFilter.get(tileData.tileNum);
+            int sum = 0;
+            for (boolean b : filterDatas) {
+                sum += b ? 1 : 0;
+            }
+            byte[] filteredByteArray = new byte[sum];
+            int filterIndex = 0;
+            int basecallIndex = 0;
+            for (boolean filterData : filterDatas) {
+                byte readByte = unNibbledByteArray[filterIndex];
+                if (filterData) {
+                    filteredByteArray[basecallIndex] = readByte;
+                    basecallIndex++;
                 }
+                filterIndex++;
             }
-            uncompressedFilteredByteBuffer.flip();
-            cachedTile[totalCycleCount] = uncompressedFilteredByteBuffer.slice();
+            cachedTile[totalCycleCount] = filteredByteArray;
         } else {
-            cachedTile[totalCycleCount] = unNibbledByteBuffer;
+            cachedTile[totalCycleCount] = unNibbledByteArray;
         }
+        cachedTilePosition[totalCycleCount] = 0;
+    }
+
+    public CycleData[] getCycleData() {
+        return cycleData;
+    }
+
+    public int getHeaderSize() {
+        return headerSize;
+    }
+
+    public List<File> getFilesForCycle(int i) {
```

Comment:
The way this is used from `CheckIlluminaDirectory.doWork()` is kind of awkward. That code calls `getAllTiles()` and then uses the map key to call this method. If instead `getAllTiles()` return a list of pairs of `List<TitleData>` and `List<File>` then I think you'd only need a single API. Actually it looks like the entire block in `CheckIlluminaDirectory` could be moved into this class, as `validate()` or some similar method.